app/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -0,0 +1,228 @@
+
+    @Volatile
+    private var isReceivingMessages = false
+
+    init {
+        awalaSetupJob = awalaScope.launch {
+            Awala.setUp(context)
+            checkIfAwalaAppInstalled()
+            isAwalaSetUp = true
+            awalaSetupJob = null
+        }
+    }
+
+    override suspend fun sendMessage(
+        message: Message,
+        recipient: MessageRecipient,
+    ) {
+        val firstPartyEndpointNodeId = awalaRepository.getServerFirstPartyEndpointNodeId()
+            ?: registerFirstPartyEndpointIfNeeded()?.nodeId
+            ?: ""
+
+        val thirdPartyEndpointNodeId = when (recipient) {
+            is MessageRecipient.Server -> {
+                recipient.nodeId
+                    ?: awalaRepository.getServerThirdPartyEndpointNodeId()
+                    ?: importServerThirdPartyEndpointIfNeeded()?.nodeId
+                    ?: ""
+            }
+            else -> {
+                throw IllegalStateException("User messages are not supported yet!")
+            }
+        }
+
+        if (firstPartyEndpointNodeId.isEmpty() || thirdPartyEndpointNodeId.isEmpty()) {
+            throw IllegalStateException("some of ids is empty: $firstPartyEndpointNodeId ; $thirdPartyEndpointNodeId")
+        }
+
+        Log.d(TAG, "sendMessage() from $firstPartyEndpointNodeId to $thirdPartyEndpointNodeId: ${message})")
+
+        val firstPartyEndpoint = loadNonNullFirstPartyEndpoint(firstPartyEndpointNodeId)
+        val thirdPartyEndpoint = loadNonNullThirdPartyEndpoint(thirdPartyEndpointNodeId)
+
+        GatewayClient.sendMessage(
+            OutgoingMessage.build(
+                type = message.type.value,
+                content = message.content,
+                senderEndpoint = firstPartyEndpoint,
+                recipientEndpoint = thirdPartyEndpoint,
+            )
+        )
+    }
+
+    override suspend fun isAwalaInstalled(currentScreen: Route): Boolean {
+        if (!isAwalaSetUp) {
+            awalaSetupJob?.join()
+        }
+        return if (currentScreen == Route.AwalaNotInstalled) {
+            checkIfAwalaAppInstalled()
+        } else {
+            isAwalaInstalledOnDevice ?: checkIfAwalaAppInstalled()
+        }
+    }
+
+    private suspend fun startReceivingMessages() {
+        if (isReceivingMessages) {
+            return
+        }
+
+        awalaScope.launch {
+            Log.d(TAG, "start receiving messages...")
+            GatewayClient.receiveMessages().collect { message ->
+                // TODO Remove first message.ack() before publishing the app.
+                // It's here to avoid the server getting stuck with messages that can't be processed.
+                message.ack()
+                _messages.emit(
+                    Message(
+                        type = MessageType.from(message.type),
+                        content = message.content,
+                    ).also { Log.d(TAG, "Receive message: ($it)") }
+                )
+                message.ack()
```

Comment:
I missed this in the previous review, but this is problematic: We're acknowledging the message before it's either processed or stored for processing later. So if we fail to process that message, it will be lost forever because the gateway will have deleted its copy.

This is why the original version of this code processed the message in-band. If we find that processing can take too long, we can create some kind of queue... But I'd rather cross that bridge when we get there.

Reply:
I'm sorry, but I didn't get the problem, why a message should be lost?

Reply:
BTW I got what you meant. I think we can use Channels instead of SharedFlow, because it ensures that no items will be lost

Reply:
For example, if the app crashes whilst processing the message.

Reply:
Okay, we can keep the message-parsing inside of AwalaManager, just need to refactor it a little bit then.

Reply:
Parsing the message is just the first step of processing it.

This acknowledgement is core to the architecture of Awala: It must only be called once it's safe for the server to delete it. If we acknowledge it after parsing, but then the app immediately crashes (for example), we'll have lost the message forever. We can't afford to lose messages in this environment, where there could be life critical information exchanged.

This is why I'm saying that can't ack it "before it's either processed or stored for processing later". If we **really** have to retrieve and ack messages fast, then the only option is to create some kind of background queue (which stores incoming messages on disk and then another component processes them asynchronously). 

However, I think this would be an overkill at this point, unless you've done some performance benchmarking and have reason to believe that the current version would cause issues?

Reply:
I got the problem, thanks.

But if you're talking about current version of `GatewayRepository` it has exactly the same problem. So I don't get how this solves our problem. The only solution is indeed to save messages at some device storage and delete them only after successful handling. If we're ready to invest time to realise this, then it can be realised :)

Reply:
But as from my side, we're talking about the worst-case scenario, when the app would crash during message processing, which is close to unreal scenario (why it should crash?)

Reply:
> But if you're talking about current version of GatewayRepository it has exactly the same problem

If you're talking about the `ack()` call I asked you to remove (and you removed), that's just a temporary thing the other developer forgot to remove.

Apart from that, as I understand the current version of the code this branch was based on, it does all the processing in band and then calls `ack()` at the end. So it isn't susceptible to losing messages.

(Having said this, this is what the other developer was supposed to do, but the original `GatewayRepository` is almost impossible to understand, so maybe it doesn't actually behave as it's supposed to)

> The only solution is indeed to save messages at some device storage and delete them only after successful handling. If we're ready to invest time to realise this, then it can be realised :)

But again, I'm saying let's not do that unless we really have to, and I'm not seeing any reason to do it at this point.

I'm just saying we should do all the processing in-band for now, and only move to a queue if we ever experience issues. [Awala Ping](https://play.google.com/store/apps/details?id=tech.relaycorp.ping) does all the processing in-band and we haven't had any issues to date, though admittedly that's a much simpler app.

> But as from my side, we're talking about the worst-case scenario, when the app would crash during message processing, which is close to unreal scenario (why it should crash?)

A bug in our code, Android killed it, the battery dies, etc.

Reply:
Note that we can still move the in-band processing to another thread if you want to. All I'm saying is that `ack()` should only ever be called when it's safe for the gateway to delete the message.

Reply:
> Apart from that, as I understand the current version of the code this branch was based on, it does all the processing in band and then calls ack() at the end. So it isn't susceptible to losing messages.

Not exactly. It does only parsing in-band and then pass the data to flow, which has a subscriber, which itself will process the message content. So the only difference is that in the `GatewayRepository` parsing is performed in band, not message handling. I sketched some solution this morning to preserve the flexibility of the `AwalaManager` and parse messages in-band, to not crash at least on the stage of parsing.

Reply:
> > Apart from that, as I understand the current version of the code this branch was based on, it does all the processing in band and then calls ack() at the end. So it isn't susceptible to losing messages.
> 
> Not exactly. It does only parsing in-band and then pass the data to flow, which has a subscriber, which itself will process the message content. 

Oh, yeah, that's true 😞 

> I sketched some solution this morning to preserve the flexibility of the AwalaManager and parse messages in-band, to not crash at least on the stage of parsing

It's still not reliable enough to go live with it. Should I create an issue on Jira to fix it later, or do you want to do it in this PR?

Reply:
I would do it not in this PR. Let's skip it for now, and I can implement it while you're dealing with this issue with message delivering